### PR TITLE
waterfacilityrestoration model

### PR DESCRIPTION
**Pull request by longshuicy**
Ready to be reviewed.

To test, run the `test_waterfacilityrestoration.py`

I wrote two utility function to extract the inventory_class from mapping rules; for both new format vs old format. 
Although such nested rule might not happen for restorations, but the utility function can parse it. E.g
With some "fun" recursive trick, `{"AND": ["java.lang.String utilfcltyc EQUALS 'PWT2'", {"OR": ["java.lang.String utilfcltyc EQUALS 'PPP2'", "java.lang.String utilfcltyc EQUALS 'PPP1'"]}]}` -->` PWT2+PPP2/PPP1`

If you want to test with local restorations, use this blob in the test
```
# use local mapping and restoration curves
    # entry_1 = {"Restoration ID Code": RestorationCurveSet.from_json_file("/tests/data/restorationset.json")}
    # # capable of handling multiple rules
    # rules_1 = [["java.lang.String utilfcltyc EQUALS 'PWT2'"]]
    # # rules_1 = [["java.lang.String utilfcltyc EQUALS 'PWT2'", "java.lang.String utilfcltyc EQUALS 'PPP2'"]]
    # # rules_1 = {"AND": ["java.lang.String utilfcltyc EQUALS 'PWT2'", "java.lang.String utilfcltyc EQUALS 'PPP2'"]}
    # # rules_1 = {"AND": ["java.lang.String utilfcltyc EQUALS 'PWT2'", {"OR": ["java.lang.String utilfcltyc EQUALS "
    # #                                                                  "'PPP2'", "java.lang.String utilfcltyc EQUALS "
    # #                                                                  "'PPP1'"]}]}
    # mapping_1 = Mapping(entry_1, rules_1)
    # mapping_set = {
    #     'id': 'restoration mapping id',
    #     'name': 'testing local mapping object creation',
    #     'hazardType': 'earthquake',
    #     'inventoryType': 'water_facility',
    #     'mappings': [
    #         mapping_1,
    #     ],
    #     'mappingType': 'restoration'
    # }
    # local_mapping_set = MappingSet(mapping_set)
    # wf_rest.set_input_dataset('dfr3_mapping_set', local_mapping_set)
```

### CODE REVIEW

File: pyincore/dfr3service.py [hunk 1]
```diff
@@ -364,6 +364,53 @@ def _eval_criterion(rule, properties):
...
+                return "NA"
+            else:
+                inventory_class = ""
+                for i, and_rules in enumerate(rules):
+                    # or
+                    if i != 0:
+                        inventory_class += "/"
+
+                    for j, rule in enumerate(and_rules):
+                        # and
```
Comment (diegoac2) [inline pyincore/dfr3service.py:381]: what does this comment mean?

File: pyincore/dfr3service.py [hunk 2]
```diff
@@ -364,6 +364,53 @@ def _eval_criterion(rule, properties):
...
+    @staticmethod
+    def extract_inventory_class_legacy(rules):
+        # legacy
+        if isinstance(rules, list):
+            if rules == [[]] or rules == [] or rules == [None]:
+                return "NA"
+            else:
+                inventory_class = ""
+                for i, and_rules in enumerate(rules):
+                    # or
```
Comment (diegoac2) [inline pyincore/dfr3service.py:376]: what does this comment mean?

File: pyincore/dfr3service.py [hunk 3]
```diff
@@ -364,6 +364,53 @@ def _eval_criterion(rule, properties):
 
         return matched
 
+    @staticmethod
+    def extract_inventory_class_legacy(rules):
+        # legacy
```
Comment (diegoac2) [inline pyincore/dfr3service.py:369]: this seems like an unnecessary comment

File: pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        if end_time is None:
+            end_time = 365.0
+
+        time_interval = self.get_parameter("time_interval")
+        if time_interval is None:
+            time_interval = 1  # 1 day
+
+        pf_interval = self.get_parameter("pf_interval")
+        if pf_interval is None:
+            pf_interval = 0.1  # 0.1
```
Comment (diegoac2) [inline pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:49]: this comment seems irrelevant

File: pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    def run(self):
+        """Performs Water facility restoration analysis by using the parameters from the spec
+        and creates an output dataset in csv format
+
+        Returns:
+            bool: True if successful, False otherwise
+        """
+        mapping_set = self.get_input_dataset("dfr3_mapping_set")
+
+        # Get Fragility key
```
Comment (diegoac2) [inline pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:34]: this comment seems redundant

File: pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        if restoration_key is None:
+            restoration_key = "Restoration ID Code"
+
+        end_time = self.get_parameter("end_time")
+        if end_time is None:
+            end_time = 365.0
+
+        time_interval = self.get_parameter("time_interval")
+        if time_interval is None:
+            time_interval = 1  # 1 day
```
Comment (diegoac2) [inline pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:45]: maybe rename this variable so it's more clear that the time interval is in days? If not, the comment doesn't really make the code more readable

Comment (longshuicy) [reply pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:45]: oh maybe i'll put the day in the input parameter description section in the spec?
this could be used for any time interval; it's just a convention that it's using day

File: pyincore/dfr3service.py [hunk 7]
```diff
@@ -364,6 +364,53 @@ def _eval_criterion(rule, properties):
...
+                    for j, rule in enumerate(and_rules):
+                        # and
+                        if j != 0:
+                            inventory_class += "+"
+                        inventory_class += rule.split(" ")[3].strip('\'').strip('\"')
+                return inventory_class
+
+    @staticmethod
+    def extract_inventory_class(rules):
+        # new format
```
[the remark this answers is not in the document]
Comment (longshuicy) [reply pyincore/dfr3service.py:389]: removed

**Comment by navarroc** [comment]
With the defaults in the test, for the percent functionality, I'm getting some inf values. Do we need to check for that case?
e.g. 
"PPP2","1.0","inf","inf","inf","inf"
"PWE1","0.0","-inf","-inf","-inf","-inf"

**Comment by longshuicy** [comment]
> With the defaults in the test, for the percent functionality, I'm getting some inf values. Do we need to check for that case? e.g. "PPP2","1.0","inf","inf","inf","inf" "PWE1","0.0","-inf","-inf","-inf","-inf"

I think inf is expected? cuz it won't actually reaching 1.0 or 0.0?

**Comment by longshuicy** [comment]
> > With the defaults in the test, for the percent functionality, I'm getting some inf values. Do we need to check for that case? e.g. "PPP2","1.0","inf","inf","inf","inf" "PWE1","0.0","-inf","-inf","-inf","-inf"
> 
> I think inf is expected? cuz it won't actually reaching 1.0 or 0.0?

maybe i should start from 0.01 and ends at 0.99?

**Review by vnarah2 — COMMENTED** [review]
@longshuicy what does this rule mean, in a sentence? Is it just checking if facility type is PWT2/PPP2/PPP1? I don't follow how AND/OR conditions can apply for a single column

```
{"AND": ["java.lang.String utilfcltyc EQUALS 'PWT2'", {"OR": ["java.lang.String utilfcltyc EQUALS 'PPP2'", "java.lang.String utilfcltyc EQUALS 'PPP1'"]}]}
```

**Comment by longshuicy** [comment]
> @longshuicy what does this rule mean, in a sentence? Is it just checking if facility type is PWT2/PPP2/PPP1? I don't follow how AND/OR conditions can apply for a single column
> 
> ```
> {"AND": ["java.lang.String utilfcltyc EQUALS 'PWT2'", {"OR": ["java.lang.String utilfcltyc EQUALS 'PPP2'", "java.lang.String utilfcltyc EQUALS 'PPP1'"]}]}
> ```

The class component is extracted based on the mapping rules; usually it will only have one rule
say `PWT2` --> restoration curve x

but just in case if people writes complicated rules
say `PWT2 and PPP2` --> restoration curve x
this is still able to capture that information. the class component then will become `PWT2+PPP2`

File: tests/pyincore/analyses/waterfacilityrestoration/test_waterfacilityrestoration.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+def run_with_base_class():
+    client = IncoreClient(pyglobals.INCORE_API_DEV_URL)
+    wf_rest = WaterFacilityRestoration(client)
+
+    # Load restoration mapping
+    restorationsvc = RestorationService(client)
+    mapping_set = MappingSet(restorationsvc.get_mapping("61f075ee903e515036cee0a5"))  # new format of mapping
+    wf_rest.set_input_dataset('dfr3_mapping_set', mapping_set)
+    wf_rest.set_parameter("result_name", "wf_restoration.csv")
```
Comment (vnarah2) [inline tests/pyincore/analyses/waterfacilityrestoration/test_waterfacilityrestoration.py:20]: Should there be a .csv here? I think we just use the name without extension

File: pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                {
+                    'id': 'result_name',
+                    'required': True,
+                    'description': 'result dataset name',
+                    'type': str
+                },
+                {
+                    'id': 'end_time',
+                    'required': False,
+                    'description': 'end time. Default to 365.',
```
Comment (vnarah2) [inline pyincore/analyses/waterfacilityrestoration/waterfacilityrestoration.py:138]: can you mention, end time in days?

File: tests/pyincore/analyses/waterfacilityrestoration/test_waterfacilityrestoration.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    wf_rest.set_parameter("restoration_key", "Restoration ID Code")
+    wf_rest.set_parameter("end_time", 100.0)
+    wf_rest.set_parameter("time_interval", 1.0)
+    wf_rest.set_parameter("pf_interval", 0.01)
```
Comment (vnarah2) [inline tests/pyincore/analyses/waterfacilityrestoration/test_waterfacilityrestoration.py:24]: feel free to ignore this, I think a better default would be 0.05 that way there are 20 rows for each type, instead of 100

**Review by vnarah2 — APPROVED** [review]

**Review by diegoac2 — APPROVED** [review]
